### PR TITLE
haskell.compiler.ghc961: init at 9.6.1

### PR DESCRIPTION
**Pull request by sternenseemann**
xhtml seems to be built unconditionally now which is at least one thing improved by hadrian.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/ghc/9.6.1.nix [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import ./common-hadrian.nix rec {
+  version = "9.6.0.20230111";
+  url = "https://downloads.haskell.org/~ghc/9.6.1-alpha1/ghc-${version}-src.tar.xz";
+  sha256 = "c9647f22d8858d43c4a654ae1ef4b94660337f5658417d6344dca0d3946c6945";
+}
```
Comment (maralorn) [inline pkgs/development/compilers/ghc/9.6.1.nix:5]: This is quite beautiful.

**Comment by guibou** [comment]
Thank you for this work.

I've tested it and the compilers works perfectly. However, `ghcWithPackages` is not working.

Initially, it fails because of:

```
ghc> rm: cannot remove '/nix/store/jfk7c17xgyfvxjbf9cg3mmcwc9rwh5sm-ghc-9.6.0.20230128-with-packages/lib/ghc-9.6.0.20230128/package.conf.d/package.cache.lock': No such file or directory
```

Looks like the `package.cache.lock` file does not exists anymore. I tried to "fix" in a backward compatibly way the `rm` call by adding `-f`:

```diff
diff --git a/pkgs/development/haskell-modules/with-packages-wrapper.nix b/pkgs/development/haskell-modules/with-packages-wrapper.nix
index 77ccd301b43..9aa326a8b5b 100644
--- a/pkgs/development/haskell-modules/with-packages-wrapper.nix
+++ b/pkgs/development/haskell-modules/with-packages-wrapper.nix
@@ -156,8 +156,8 @@ symlinkJoin {
      # to another nix derivation, so they are not writable.  Removing
      # them allow the correct behavior of ghc-pkg recache
      # See: https://github.com/NixOS/nixpkgs/issues/79441
-     rm ${packageCfgDir}/package.cache.lock
-     rm ${packageCfgDir}/package.cache
+     rm -f ${packageCfgDir}/package.cache.lock
+     rm -f ${packageCfgDir}/package.cache
 
      $out/bin/${ghcCommand}-pkg recache
      ''}
```

But now it fails with:

```
hc> WARNING: settings file doesn't exist "/nix/store/1mf5mh65drkwpp77l8l2a0h0rl8d3h1d-ghc-9.6.0.20230128-with-packages/lib/ghc-9.6.0.20230128/settings"
ghc> cannot know target platform so guessing target == host (native compiler).
ghc> WARNING: settings file doesn't exist "/nix/store/1mf5mh65drkwpp77l8l2a0h0rl8d3h1d-ghc-9.6.0.20230128-with-packages/lib/ghc-9.6.0.20230128/settings"
ghc> cannot know target platform so guessing target == host (native compiler).
ghc> There are problems in package hspec-2.9.7:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc> There are problems in package HUnit-1.6.2.0:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "deepseq-1.4.8.0" doesn't exist
ghc> There are problems in package temporary-1.3:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "directory-1.3.8.0" doesn't exist
ghc>   dependency "exceptions-0.10.7" doesn't exist
ghc>   dependency "filepath-1.4.100.0" doesn't exist
ghc>   dependency "transformers-0.6.0.4" doesn't exist
ghc>   dependency "unix-2.8.0.0" doesn't exist
ghc> There are problems in package hspec-core-2.9.7:
ghc>   dependency "array-0.5.4.0" doesn't exist
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "deepseq-1.4.8.0" doesn't exist
ghc>   dependency "directory-1.3.8.0" doesn't exist
ghc>   dependency "filepath-1.4.100.0" doesn't exist
ghc>   dependency "ghc-9.6.0.20230128" doesn't exist
ghc>   dependency "ghc-boot-th-9.6.0.20230128" doesn't exist
ghc>   dependency "stm-2.5.1.0" doesn't exist
ghc>   dependency "transformers-0.6.0.4" doesn't exist
ghc> There are problems in package hspec-discover-2.9.7:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "directory-1.3.8.0" doesn't exist
ghc>   dependency "filepath-1.4.100.0" doesn't exist
ghc> There are problems in package hspec-expectations-0.8.2:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc> There are problems in package QuickCheck-2.14.2:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "containers-0.6.6" doesn't exist
ghc>   dependency "deepseq-1.4.8.0" doesn't exist
ghc>   dependency "template-haskell-2.20.0.0" doesn't exist
ghc>   dependency "transformers-0.6.0.4" doesn't exist
ghc> There are problems in package call-stack-0.4.0:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc> There are problems in package random-1.2.1.1:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "bytestring-0.11.4.0" doesn't exist
ghc>   dependency "deepseq-1.4.8.0" doesn't exist
ghc>   dependency "mtl-2.3.1" doesn't exist
ghc> There are problems in package ansi-terminal-0.11.4:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc> There are problems in package clock-0.8.3:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc> There are problems in package quickcheck-io-0.2.0:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc> There are problems in package setenv-0.1.1.3:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "unix-2.8.0.0" doesn't exist
ghc> There are problems in package tf-random-0.5:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "time-1.12.2" doesn't exist
ghc> There are problems in package splitmix-0.1.0.4:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "deepseq-1.4.8.0" doesn't exist
ghc> There are problems in package colour-2.3.6:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc> There are problems in package primitive-0.7.4.0:
ghc>   dependency "base-4.18.0.0" doesn't exist
ghc>   dependency "deepseq-1.4.8.0" doesn't exist
ghc>   dependency "template-haskell-2.20.0.0" doesn't exist
ghc>   dependency "transformers-0.6.0.4" doesn't exist
ghc> The following packages are broken, either because they have a problem
ghc> listed above, or because they depend on a broken package.
ghc> hspec-2.9.7
ghc> HUnit-1.6.2.0
ghc> temporary-1.3
ghc> hspec-core-2.9.7
ghc> hspec-discover-2.9.7
ghc> hspec-expectations-0.8.2
ghc> QuickCheck-2.14.2
ghc> call-stack-0.4.0
ghc> random-1.2.1.1
ghc> ansi-terminal-0.11.4
ghc> clock-0.8.3
ghc> quickcheck-io-0.2.0
ghc> setenv-0.1.1.3
ghc> tf-random-0.5
ghc> splitmix-0.1.0.4
ghc> colour-2.3.6
ghc> primitive-0.7.4.0
```

(that's for one specific package db, the error may vary depending on the `ghcWithPackages` call).

I'll try to investigate later, I wanted to report this breakage, but this does not block the merge of the compiler from my point of view.

**Comment by sternenseemann** [comment]
@guibou That is fixed in #212107, this branch is just not based on a new enough haskell-updates. When rebased, that issue should go away (you can probably try yourself easily).

File: pkgs/development/haskell-modules/configuration-ghc-9.6.x.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  hpc = null;
+  integer-gmp = null;
+  libiserv = null;
+  mtl = null;
+  parsec = null;
+  pretty = null;
+  process = null;
+  rts = null;
+  stm = null;
+  system-cxx-std-lib = null;
```
Comment (sternenseemann) [inline pkgs/development/haskell-modules/configuration-ghc-9.6.x.nix:41]: need to check if this makes sense (also `rts` and `hpc`) – if they are not on hackage it's probably sensible to remove them here.

**Comment by guibou** [comment]
I've rebased and applied the following diff:

```diff
diff --git a/pkgs/development/compilers/ghc/9.6.1.nix b/pkgs/development/compilers/ghc/9.6.1.nix
index f9c9a4b1e6a..574ff88cf46 100644
--- a/pkgs/development/compilers/ghc/9.6.1.nix
+++ b/pkgs/development/compilers/ghc/9.6.1.nix
@@ -1,5 +1,5 @@
 import ./common-hadrian.nix rec {
-  version = "9.6.0.20230128";
-  url = "https://downloads.haskell.org/~ghc/9.6.1-alpha2/ghc-${version}-src.tar.xz";
-  sha256 = "531f58dc0e352987a637095c18252e5d6fef260969eb9cb3d8447888b0922da2";
+  version = "9.6.0.20230210";
+  url = "https://downloads.haskell.org/~ghc/9.6.1-alpha3/ghc-${version}-src.tar.xz";
+  sha256 = "sha256-3vSnSFFiJm6R70YA38yoQki+1kFHF91+66za2WogT0c=";
 }
```

For alpha 3 support.

**Comment by sternenseemann** [comment]
@guibou Thanks, did so on my branch as well. Also tested `ghcWithPackages` which seems to work after rebasing.

**Pull request merged**